sotodlib/site_pipeline/ml_mapmaking.py

```diff
@@ -0,0 +1,312 @@
+from argparse import ArgumentParser
+import numpy as np, sys, time, warnings, os, so3g
+from sotodlib.core import Context, AxisManager, IndexAxis
+from sotodlib import mapmaking
+from sotodlib.io import metadata   # PerDetectorHdf5 work-around
+from sotodlib import tod_ops
+from sotodlib.tod_ops import filters
+from pixell import enmap, utils, fft, bunch, wcsutils, mpi
+#import moby2.analysis.socompat
+import yaml
+
+def get_parser(parser=None):
+    if parser is None:
+        parser = ArgumentParser()
+    parser.add_argument("--config-file", type=str, default=None, 
+                     help="Path to mapmaker config.yaml file")
+
+    parser.add_argument("--query", type=str)
+    parser.add_argument("--freq", type=str)
+    parser.add_argument("--area", type=str)
+    parser.add_argument("--odir", type=str)
+    parser.add_argument("--prefix", type=str)
+    parser.add_argument('-C', "--comps",   type=str, help="T,Q, and/or U")
+    parser.add_argument("-c", "--context", type=str,  help="Context containing TODs")
+    parser.add_argument("-n", "--ntod",    type=int)
+    parser.add_argument(      "--tods",    type=str)
+    parser.add_argument(      "--nset",    type=int)
+    parser.add_argument("-N", "--nmat",    type=str)
+    parser.add_argument(      "--max-dets",type=int)
+    parser.add_argument("-S", "--site",    type=str)
+    parser.add_argument("-v", "--verbose", action="count")
+    parser.add_argument("-q", "--quiet",   action="count")
+    parser.add_argument("-@", "--center-at", type=str)
+    parser.add_argument("-w", "--window",  type=float)
+    parser.add_argument("-i", "--inject",  type=str)
+    parser.add_argument(      "--nocal",   action="store_true", default=True)
+    parser.add_argument(      "--nmat-dir", type=str)
+    parser.add_argument(      "--nmat-mode", type=str, help="How to build the noise matrix. 'build': Always build from tod. 'cache': Use if available in nmat-dir, otherwise build and save. 'load': Load from nmat-dir, error if missing. 'save': Build from tod and save.")
+    parser.add_argument("-d", "--downsample", type=int, help="Downsample TOD by this factor")
+    parser.add_argument(      "--maxiter",    type=int)
+    parser.add_argument("-T", "--tiled"  ,    type=int)
+    parser.add_argument("-W", "--wafer"  ,   type=str, nargs='+', help="Detector wafer subset to map with")
+    return parser
+
+
+def _get_config(config_file):
+    return yaml.safe_load(open(config_file,'r'))
+
+
+
+def main(config_file=None, **args):
+    cfg = _get_config(config_file)
+    
+    # Certain fields are required. Check if they are all supplied here
+    required_fields = ['query','freq','area','odir','prefix']
+    for req in required_fields:
+        if req not in cfg.keys():
+            raise KeyError("Please supply a {}".format(req))
```

Comment:
It looks to me like:
- the config_file must exist (so can't use cli only?)
- the "required_fields" must all be specified in the config file (I can't specify --freq on command line instead?)

I think those should both be loosened.  At the end of the process of reading a config file, and processing command line args ... check that all the things you need are defined.

That error message is a bit vague -- supply how?
-